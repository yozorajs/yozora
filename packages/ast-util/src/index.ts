export * from './ast/clone'
export * from './ast/collect'
export * from './ast/replace-post-order'
export * from './ast/replace-pre-order'
export * from './ast/search'
export * from './ast/toc'
export * from './ast/traverse'
export * from './ast/util'
export * from './definition'
export * from './excerpt'
export * from './footnote-definition'
export * from './url'
