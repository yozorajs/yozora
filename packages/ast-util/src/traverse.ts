import type { Root, YastNode, YastNodeType, YastParent } from '@yozora/ast'

/**
 * Traverse yozora AST, and provide an opportunity to perform an action on
 * visited node.
 *
 * Note that the root node will not be traversed, that is, the root node will
 * never be passed into the `mutate` function..
 *
 * @param root
 * @param mutate
 * @param aimTypes
 */
export function traverseAST(
  root: Root,
  mutate: (node: YastNode, parent: YastParent, childIndex: number) => void,
  aimTypes: ReadonlyArray<YastNodeType> | null = null,
): void {
  const visit = (u: YastNode): void => {
    const { children } = u as YastParent

    // Recursively visit.
    if (children != null) {
      for (let i = 0; i < children.length; ++i) {
        const v = children[i]
        if (aimTypes == null || aimTypes.indexOf(v.type) > -1) {
          mutate(v, (u as unknown) as YastParent, i)
        }
        visit(v)
      }
    }
  }
  visit(root)
}
