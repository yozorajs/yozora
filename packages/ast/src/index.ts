export * from './ast'
export * from './nodes/blockquote'
export * from './nodes/break'
export * from './nodes/code'
export * from './nodes/definition'
export * from './nodes/delete'
export * from './nodes/emphasis'
export * from './nodes/footnote'
export * from './nodes/footnote-definition'
export * from './nodes/footnote-reference'
export * from './nodes/frontmatter'
export * from './nodes/heading'
export * from './nodes/html'
export * from './nodes/image-reference'
export * from './nodes/image'
export * from './nodes/inline-code'
export * from './nodes/inline-math'
export * from './nodes/link-reference'
export * from './nodes/link'
export * from './nodes/list-item'
export * from './nodes/list'
export * from './nodes/math'
export * from './nodes/paragraph'
export * from './nodes/root'
export * from './nodes/strong'
export * from './nodes/table-cell'
export * from './nodes/table-row'
export * from './nodes/table'
export * from './nodes/text'
export * from './nodes/thematic-break'
