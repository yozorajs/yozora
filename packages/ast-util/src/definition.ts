import type {
  Definition,
  Root,
  YastAssociation,
  YastNodeType,
} from '@yozora/ast'
import { DefinitionType } from '@yozora/ast'
import { traverseAST } from './ast/traverse'

/**
 * Calc definition map from Yozora AST.
 * @param root
 * @param aimTypes
 * @returns
 */
export function collectIdentifiers(
  root: Readonly<Root>,
  aimTypes: ReadonlyArray<YastNodeType>,
  presetDefinitions: ReadonlyArray<Readonly<YastAssociation>> = [],
): Record<string, true> {
  const identifiers: Record<string, true> = {}

  // Traverse Yozora AST and collect identifier of definitions.
  traverseAST(root, aimTypes, (node): void => {
    const definition = (node as unknown) as YastAssociation
    identifiers[definition.identifier] = true
  })

  // Collect from preset definitions.
  for (const definition of presetDefinitions) {
    identifiers[definition.identifier] = true
  }

  return identifiers
}

/**
 * Calc definition map from Yozora AST.
 * @param root
 * @param aimTypes
 * @returns
 */
export function collectDefinitions(
  root: Readonly<Root>,
  aimTypes: ReadonlyArray<YastNodeType> = [DefinitionType],
  presetDefinitions: ReadonlyArray<Readonly<Omit<Definition, 'type'>>> = [],
): Record<string, Readonly<Definition>> {
  const definitions: Record<string, Readonly<Definition>> = {}
  const add = (definition: Readonly<Definition>): void => {
    const { identifier } = definition

    /**
     * If there are several matching definitions, the first one takes precedence
     * @see https://github.github.com/gfm/#example-173
     */
    if (definitions[identifier] != null) return
    definitions[identifier] = definition
  }

  // Traverse Yozora AST and collect definitions.
  traverseAST(root, aimTypes, (node): void => {
    const definition = node as Definition
    add(definition)
  })

  // Add preset definitions at the end to avoid incorrectly overwriting custom
  // definitions defined in the Yozora AST.
  for (const definition of presetDefinitions) {
    add({ type: DefinitionType, ...definition })
  }

  return definitions
}
