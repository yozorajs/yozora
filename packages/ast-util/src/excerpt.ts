import type { Literal, Parent, Root } from '@yozora/ast'
import { TextType } from '@yozora/ast'
import { shallowCloneAst } from './ast/clone'
import { searchNode } from './ast/search'

/**
 * Calc excerpt ast from the original ast.
 */
export function calcExcerptAst(immutableRoot: Readonly<Root>, pruneLength: number): Readonly<Root> {
  if (immutableRoot.children.length <= 0) return immutableRoot

  let totalExcerptLengthSoFar = 0
  let parentOfLastLiteralNode: Parent | null = null as any
  let indexOfLastLiteralNode = 0
  const excerptAst = shallowCloneAst(immutableRoot, (node, parent, index) => {
    if (totalExcerptLengthSoFar >= pruneLength) return true
    const { value } = node as Literal
    if (value != null) {
      parentOfLastLiteralNode = parent
      indexOfLastLiteralNode = index
      totalExcerptLengthSoFar += value.length
    }
    return false
  })

  if (
    parentOfLastLiteralNode !== null &&
    parentOfLastLiteralNode.type === TextType &&
    totalExcerptLengthSoFar > pruneLength
  ) {
    // Try truncate last LiteralNode
    const parent = parentOfLastLiteralNode!
    parent.children = parent.children.map((node, i) => {
      if (i !== indexOfLastLiteralNode) return node
      return {
        ...node,
        value: (node as Literal).value.slice(totalExcerptLengthSoFar - pruneLength),
      }
    })
  }
  return excerptAst
}

/**
 * Get excerpt ast from the full AST, if there is a matched excerptSeparator, then prune it until
 * meet the separator, otherwise, use the `calcExcerptAst` to generate the excerpt.
 * @returns
 */
export function getExcerptAst(fullAst: Root, pruneLength: number, excerptSeparator?: string): Root {
  if (excerptSeparator != null) {
    const separator = excerptSeparator.trim()

    const childIndexList: number[] | null = searchNode(fullAst, node => {
      const { value } = node as Literal
      return typeof value === 'string' && value.trim() === separator
    })

    if (childIndexList != null) {
      const excerptAst = { ...fullAst }
      let node: Parent = excerptAst
      for (const childIndex of childIndexList) {
        const nextNode = { ...node.children[childIndex] } as unknown as Parent
        node.children = node.children.slice(0, childIndex)
        node.children.push(nextNode)
        node = nextNode
      }
      return excerptAst
    }
  }

  // Try to truncate excerpt.
  const excerptAst = calcExcerptAst(fullAst, pruneLength)
  return excerptAst
}
