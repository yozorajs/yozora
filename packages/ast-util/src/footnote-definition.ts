import type {
  Footnote,
  FootnoteDefinition,
  FootnoteReference,
  Paragraph,
  Root,
  YastNodeType,
} from '@yozora/ast'
import {
  FootnoteDefinitionType,
  FootnoteReferenceType,
  FootnoteType,
  ParagraphType,
} from '@yozora/ast'
import { collectNodes } from './ast/collect'
import { shallowMutateAstInPostorder } from './ast/replace-post-order'
import { traverseAst } from './ast/traverse'
import type { NodeMatcher } from './ast/util'

/**
 * Collect footnote reference definitions in a pre-order traversal.
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @returns
 */
export function collectFootnoteDefinitions(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<YastNodeType> | NodeMatcher = [
    FootnoteDefinitionType,
  ],
): FootnoteDefinition[] {
  const results: FootnoteDefinition[] = collectNodes(
    immutableRoot,
    aimTypesOrNodeMatcher,
  )
  const identifierMap: Record<string, true> = {}

  // filter duplicated footnote reference definitions with existed identifer.
  const deDuplicatedResults: FootnoteDefinition[] = results.filter(item => {
    const { identifier } = item
    if (identifierMap[identifier]) return false
    identifierMap[identifier] = true
    return true
  })

  return deDuplicatedResults
}

/**
 * Traverse yozora ast and generate a footnote reference definition map.
 *
 * Note that if `presetFootnoteDefinitions` or `preferReference` specified,
 * the input AST root may be modified.
 *
 * The `identifierPrefix` only works with `preferReferences=true`.
 *
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @param presetFootnoteDefinitions
 * @param preferReferences
 * @param identifierPrefix
 * @returns
 */
export function calcFootnoteDefinitionMap(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<YastNodeType> | NodeMatcher = [
    FootnoteDefinitionType,
  ],
  presetFootnoteDefinitions: ReadonlyArray<FootnoteDefinition> = [],
  preferReferences = false,
  identifierPrefix = 'footnote-',
): {
  root: Readonly<Root>
  footnoteDefinitionMap: Record<string, Readonly<FootnoteDefinition>>
} {
  const footnoteDefinitionMap: Record<string, Readonly<FootnoteDefinition>> = {}

  // Traverse Yozora AST and collect footnote definitions.
  traverseAst(immutableRoot, aimTypesOrNodeMatcher, (node): void => {
    const footnoteDefinition = node as FootnoteDefinition
    const { identifier } = footnoteDefinition

    /**
     * If there are several matching definitions, the first one takes precedence.
     * @see https://github.github.com/gfm/#example-173
     */
    if (footnoteDefinitionMap[identifier] != null) return
    footnoteDefinitionMap[identifier] = footnoteDefinition
  })

  /**
   * Add preset footnote reference definitions at the end to avoid incorrectly
   * overwriting custom defined footnote reference definitions.
   */
  const validPresetFootnoteDefinitions: FootnoteDefinition[] = []
  for (const footnoteDefinition of presetFootnoteDefinitions) {
    if (footnoteDefinition[footnoteDefinition.identifier] != null) continue
    footnoteDefinitionMap[footnoteDefinition.identifier] = footnoteDefinition
    validPresetFootnoteDefinitions.push(footnoteDefinition)
  }

  // Append the preset footnote reference definitions to the end of the
  // root.children after the ones generated from footnotes appended.
  let root: Root =
    validPresetFootnoteDefinitions.length > 0
      ? {
          ...immutableRoot,
          children: immutableRoot.children.concat(
            validPresetFootnoteDefinitions,
          ),
        }
      : immutableRoot

  // Replace footnotes into footnote references and footnote reference definitions.
  if (preferReferences) {
    root = replaceFootnotesInReferences(
      immutableRoot,
      footnoteDefinitionMap,
      identifierPrefix,
    )
  }

  return { root, footnoteDefinitionMap }
}

/**
 * You need to be careful because this operation is irreversible!
 *
 * Replace inline footnote with a footnote reference and a footnote reference
 * definition, the create a new footnote will be assigned a unique incremental
 * number (string), and appended it into the second param `footnoteDefinitions`
 * passed.
 *
 * @param immutableRoot
 * @param footnoteDefinitionMap
 * @param identifierPrefix
 * @returns
 */
export function replaceFootnotesInReferences(
  immutableRoot: Readonly<Root>,
  footnoteDefinitionMap: Record<string, Readonly<FootnoteDefinition>>,
  identifierPrefix = 'footnote-',
): Root {
  let footnoteId = 1
  const footnoteDefinitions: FootnoteDefinition[] = []

  const root: Root = shallowMutateAstInPostorder(
    immutableRoot,
    [FootnoteType],
    node => {
      const footnote = node as Footnote
      for (; ; footnoteId += 1) {
        const label = String(footnoteId)
        const identifier = identifierPrefix + label
        if (
          footnoteDefinitionMap[label] != null ||
          footnoteDefinitionMap[identifier] != null
        ) {
          continue
        }

        const paragraph: Paragraph = {
          type: ParagraphType,
          children: footnote.children,
        }

        const footnoteDefinition: FootnoteDefinition = {
          type: FootnoteDefinitionType,
          identifier,
          label,
          children: [paragraph],
        }

        const footnoteReference: FootnoteReference = {
          type: FootnoteReferenceType,
          label,
          identifier,
        }

        if (footnote.position != null) {
          footnoteReference.position = footnote.position
        }

        // eslint-disable-next-line no-param-reassign
        footnoteDefinitionMap[identifier] = footnoteDefinition
        footnoteDefinitions.push(footnoteDefinition)

        // Replace the inline footnote with a footnote reference,
        // the relevant created footnote reference definition will be appended to
        // the end of the root.children.
        return footnoteReference
      }
    },
  )

  // Append footnote definitions to the end of the root.children
  return footnoteDefinitions.length > 0
    ? { ...root, children: root.children.concat(footnoteDefinitions) }
    : root
}
