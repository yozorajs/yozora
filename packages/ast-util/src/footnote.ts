import type {
  Footnote,
  FootnoteDefinition,
  FootnoteReference,
  Paragraph,
  Root,
  YastNodeType,
} from '@yozora/ast'
import {
  FootnoteDefinitionType,
  FootnoteReferenceType,
  FootnoteType,
  ParagraphType,
} from '@yozora/ast'
import { collectNodes } from './ast/collect'
import { shallowMutateAstInPostorder } from './ast/replace-post-order'
import { traverseAst } from './ast/traverse'
import type { INodeMatcher } from './ast/util'

export const defaultFootnoteIdentifierPrefix = 'footnote-'

/**
 * Collect footnote reference definitions in a pre-order traversal.
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @returns
 */
export function collectFootnoteDefinitions(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<YastNodeType> | INodeMatcher = [FootnoteDefinitionType],
): FootnoteDefinition[] {
  const results: FootnoteDefinition[] = collectNodes(immutableRoot, aimTypesOrNodeMatcher)

  // filter duplicated footnote reference definitions with existed identifier.
  const existedSet: Set<string> = new Set<string>()
  return results.filter(item => {
    if (existedSet.has(item.identifier)) return false
    existedSet.add(item.identifier)
    return true
  })
}

/**
 * Traverse yozora ast and generate a footnote reference definition map.
 *
 * Note that if `presetFootnoteDefinitions` or `preferReference` specified,
 * the input AST root may be modified.
 *
 * The `identifierPrefix` only works with `preferReferences=true`.
 *
 * @param immutableRoot
 * @param aimTypesOrNodeMatcher
 * @param presetFootnoteDefinitions
 * @param preferReferences
 * @param identifierPrefix
 * @returns
 */
export function calcFootnoteDefinitionMap(
  immutableRoot: Readonly<Root>,
  aimTypesOrNodeMatcher: ReadonlyArray<YastNodeType> | INodeMatcher = [FootnoteDefinitionType],
  presetFootnoteDefinitions: ReadonlyArray<FootnoteDefinition> = [],
  preferReferences = false,
  identifierPrefix = defaultFootnoteIdentifierPrefix,
): {
  root: Readonly<Root>
  footnoteDefinitionMap: Record<string, Readonly<FootnoteDefinition>>
} {
  const footnoteDefinitionMap: Record<string, Readonly<FootnoteDefinition>> = {}

  // Traverse Yozora AST and collect footnote definitions.
  traverseAst(immutableRoot, aimTypesOrNodeMatcher, (node): void => {
    const footnoteDefinition = node as FootnoteDefinition
    const { identifier } = footnoteDefinition

    /**
     * If there are several matching definitions, the first one takes precedence.
     * @see https://github.github.com/gfm/#example-173
     */
    if (footnoteDefinitionMap[identifier] === undefined) {
      footnoteDefinitionMap[identifier] = footnoteDefinition
    }
  })

  /**
   * Add preset footnote reference definitions at the end to avoid incorrectly
   * overwriting custom defined footnote reference definitions.
   */
  const additionalDefinitions: FootnoteDefinition[] = []
  for (const footnoteDefinition of presetFootnoteDefinitions) {
    if (footnoteDefinitionMap[footnoteDefinition.identifier] === undefined) {
      footnoteDefinitionMap[footnoteDefinition.identifier] = footnoteDefinition
      additionalDefinitions.push(footnoteDefinition)
    }
  }

  // Append the preset footnote reference definitions to the end of the
  // root.children after the ones generated from footnotes appended.
  let root: Root =
    additionalDefinitions.length > 0
      ? {
          ...immutableRoot,
          children: immutableRoot.children.concat(additionalDefinitions),
        }
      : immutableRoot

  // Replace footnotes into footnote references and footnote reference definitions.
  if (preferReferences) {
    root = replaceFootnotesInReferences(root, footnoteDefinitionMap, identifierPrefix)
  }

  return { root, footnoteDefinitionMap }
}

/**
 * You need to be careful because this operation is irreversible!
 *
 * Replace inline footnote with a footnote reference and a footnote reference
 * definition, the create a new footnote will be assigned a unique incremental
 * number (string), and appended it into the second param `footnoteDefinitions`
 * passed.
 *
 * @param immutableRoot
 * @param footnoteDefinitionMap
 * @param identifierPrefix
 * @returns
 */
export function replaceFootnotesInReferences(
  immutableRoot: Readonly<Root>,
  footnoteDefinitionMap: Record<string, Readonly<FootnoteDefinition>>,
  identifierPrefix = defaultFootnoteIdentifierPrefix,
): Root {
  let footnoteId = 1
  const footnoteDefinitions: FootnoteDefinition[] = []

  const nextIdentifier = (): { label: string; identifier: string } => {
    for (; ; footnoteId += 1) {
      const label = String(footnoteId)
      const identifier = identifierPrefix + label
      if (
        footnoteDefinitionMap[label] === undefined &&
        footnoteDefinitionMap[identifier] === undefined
      ) {
        return { label, identifier }
      }
    }
  }

  const root: Root = shallowMutateAstInPostorder(immutableRoot, [FootnoteType], node => {
    const footnote = node as Footnote
    const { label, identifier } = nextIdentifier()

    const paragraph: Paragraph = {
      type: ParagraphType,
      children: footnote.children,
    }

    const footnoteDefinition: FootnoteDefinition = {
      type: FootnoteDefinitionType,
      identifier,
      label,
      children: [paragraph],
    }

    const footnoteReference: FootnoteReference = {
      type: FootnoteReferenceType,
      label,
      identifier,
    }

    if (footnote.position != null) {
      footnoteReference.position = footnote.position
    }

    // eslint-disable-next-line no-param-reassign
    footnoteDefinitionMap[identifier] = footnoteDefinition
    footnoteDefinitions.push(footnoteDefinition)

    // Replace the inline footnote with a footnote reference,
    // the relevant created footnote reference definition will be appended to
    // the end of the root.children.
    return footnoteReference
  })

  // Append footnote definitions to the end of the root.children
  return footnoteDefinitions.length > 0
    ? { ...root, children: root.children.concat(footnoteDefinitions) }
    : root
}
