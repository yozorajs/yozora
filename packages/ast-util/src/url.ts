import type { Root, YastNode, YastNodeType, YastResource } from '@yozora/ast'
import { DefinitionType, ImageType, LinkType } from '@yozora/ast'
import { traverseAST } from './ast/traverse'

/**
 * Resolve url.
 */
export type UrlResolver = (
  ...pathPieces: Array<string | null | undefined>
) => string

/**
 * Join url path with `prefix` and normalize the result.
 *
 * @param prefix
 * @param path
 * @returns
 */
export const defaultUrlResolver: UrlResolver = (...pathPieces): string => {
  const pieces: string[] = pathPieces.filter(
    (x): x is string => typeof x === 'string' && x.length > 0,
  )
  if (pieces.length <= 0) return ''

  // If the last path piece is a absolute url path, then return it directly,
  // otherwise, resolved it with previous url paths.
  const lastPiece = pieces[pieces.length - 1]
  return /^([/]|\w+:[/]{2})/.test(lastPiece)
    ? lastPiece
    : pieces
        .join('/')
        .replace(/[/]+/g, '/')
        .replace(/[/][.][/]/g, '/')
        .trim()
}

/**
 * Traverse yozora ast and resolve urls for aim nodes.
 * @param ast
 * @param resolveUrl
 * @param aimTypes
 */
export function resolveUrlsForAst(
  ast: Root,
  resolveUrl: UrlResolver = defaultUrlResolver,
  aimTypes: ReadonlyArray<YastNodeType> = [DefinitionType, LinkType, ImageType],
): void {
  traverseAST(ast, aimTypes, (node): void => {
    const o = node as YastNode & YastResource
    if (o.url != null) o.url = resolveUrl(o.url)
  })
}
