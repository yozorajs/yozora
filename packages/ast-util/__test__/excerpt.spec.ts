import type { Root } from '@yozora/ast'
import { loadJSONFixture } from 'jest.setup'
import { calcExcerptAst } from '../src'

describe('basic1', function () {
  const originalAst: Readonly<Root> = loadJSONFixture('basic1.ast.json')
  const ast: Root = loadJSONFixture('basic1.ast.json')

  test('default', function () {
    const excerptAst = calcExcerptAst(ast, 140)
    expect(ast).toEqual(originalAst)
    expect(excerptAst).not.toEqual(originalAst)
    expect(excerptAst).toMatchSnapshot()
  })

  test('excerpt-40', function () {
    const excerptAst = calcExcerptAst(ast, 40)
    expect(ast).toEqual(originalAst)
    expect(excerptAst).not.toEqual(originalAst)
    expect(excerptAst).toMatchSnapshot()
  })
})
