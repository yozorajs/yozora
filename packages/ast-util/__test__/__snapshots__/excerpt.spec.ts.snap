// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic1 excerpt-140 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "将进酒",
                },
              ],
              "depth": 3,
              "type": "heading",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "君不见黄河之水天上来，奔流到海不复回。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
君不见高堂明镜悲白发，朝如青丝暮成雪。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
人生得意须尽欢，莫使金樽空对月。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
天生我材必有用，千金散尽还复来。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
烹羊宰牛且为乐，会须一饮三百杯。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
岑夫子，丹丘生，将进酒，杯莫停。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
与君歌一曲，请君为我倾耳听。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
钟鼓馔玉不足贵，但愿长醉不复醒。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
古来圣贤皆寂寞，惟有饮者留其名。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
陈王昔时宴平乐，斗酒十千恣欢谑。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
主人何为言少钱，径须沽取对君酌。",
                },
                Object {
                  "type": "break",
                },
                Object {
                  "type": "text",
                  "value": "
五花马、千金裘，呼儿将出换美酒，与尔同销万古愁。",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "blockquote",
        },
        Object {
          "children": Array [
            Object {
              "alt": "logo",
              "identifier": "logo",
              "label": "logo",
              "referenceType": "collapsed",
              "type": "imageReference",
            },
          ],
          "type": "paragraph",
        },
      ],
      "keyword": "tip",
      "title": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 18,
              "line": 2,
              "offset": 18,
            },
            "start": Object {
              "column": 8,
              "line": 2,
              "offset": 8,
            },
          },
          "type": "text",
          "value": "将进酒 -- 李白 ",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 23,
              "line": 2,
              "offset": 23,
            },
            "start": Object {
              "column": 18,
              "line": 2,
              "offset": 18,
            },
          },
          "type": "inlineMath",
          "value": "x^2",
        },
      ],
      "type": "admonition",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "emphasis: ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "foo \\"",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "bar",
                },
              ],
              "type": "emphasis",
            },
            Object {
              "type": "text",
              "value": "\\" foo",
            },
          ],
          "type": "strong",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "autolink(extension): ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "http://example.com",
            },
          ],
          "type": "link",
          "url": "http://example.com",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "autolink: ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "foo@bar.example.com",
            },
          ],
          "type": "link",
          "url": "mailto:foo@bar.example.com",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Setext headings",
        },
      ],
      "depth": 1,
      "type": "heading",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Task List Items (extension):",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "foo",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "bar",
                    },
                  ],
                  "status": "doing",
                  "type": "listItem",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "baz",
                    },
                  ],
                  "status": "done",
                  "type": "listItem",
                },
              ],
              "marker": 45,
              "ordered": false,
              "spread": false,
              "type": "list",
            },
          ],
          "status": "done",
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "bim",
            },
          ],
          "status": "todo",
          "type": "listItem",
        },
      ],
      "marker": 45,
      "ordered": false,
      "spread": false,
      "type": "list",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "This is a blockquote",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "with multiple lines.",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "blockquote",
    },
    Object {
      "lang": "typescript",
      "meta": "",
      "type": "code",
      "value": "const a = 'a'
const b = 'b'
console.log('a + b:', a + b)
console.log('This is a typescript code snippets')
",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "This is a math block:",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "type": "math",
      "value": "t = \\\\lvert x-1 \\\\rvert  < \\\\epsilon",
    },
  ],
  "type": "root",
}
`;
